src/tests/suite/daos_dedup.c

```diff
@@ -0,0 +1,361 @@
+#include <daos_prop.h>
+
+#define assert_success(r) do {\
+	int __rc = (r); \
+	if (__rc != 0) \
+		fail_msg("Not successful!! Error code: " DF_RC, DP_RC(__rc)); \
+	} while (0)
+
+struct dedup_test_ctx {
+	/** Pool */
+	daos_handle_t		poh;
+	/** Container */
+	daos_handle_t		coh;
+	daos_cont_info_t	info;
+	uuid_t			uuid;
+	/** Object */
+	daos_handle_t		oh;
+	daos_obj_id_t		oid;
+	daos_key_t		dkey;
+	daos_iod_t		update_iod;
+	d_sg_list_t		update_sgl;
+	daos_iod_t		fetch_iod;
+	d_sg_list_t		fetch_sgl;
+	daos_recx_t		recx[4];
+};
+
+enum THRESHOLD_SETTING {
+	THRESHOLD_GREATER_THAN_DATA = 1,
+	THRESHOLD_LESS_THAN_DATA,
+};
+
+/** easily setup an iov and allocate */
+static void
+iov_alloc(d_iov_t *iov, size_t len)
+{
+	D_ALLOC(iov->iov_buf, len);
+	iov->iov_buf_len = iov->iov_len = len;
+}
+
+static void
+iov_alloc_str(d_iov_t *iov, const char *str)
+{
+	iov_alloc(iov, strlen(str) + 1);
+	strcpy(iov->iov_buf, str);
+}
+
+static void
+setup_from_test_args(struct dedup_test_ctx *ctx, test_arg_t *state)
+{
+	ctx->poh = state->pool.poh;
+}
+
+static void
+setup_sgl(struct dedup_test_ctx *ctx)
+{
+	dts_sgl_init_with_strings_repeat(&ctx->update_sgl, 1000, 1,
+					 "Lorem ipsum dolor sit "
+					 "amet, consectetur adipiscing elit,"
+					 " sed do eiusmod tempor incididunt ut"
+					 " labore et dolore magna aliqua.");
+
+	daos_sgl_init(&ctx->fetch_sgl, 1);
+	iov_alloc(&ctx->fetch_sgl.sg_iovs[0],
+		daos_sgl_buf_size(&ctx->update_sgl));
+}
+
+static void
+setup_keys(struct dedup_test_ctx *ctx)
+{
+	iov_alloc_str(&ctx->dkey, "dkey");
+	iov_alloc_str(&ctx->update_iod.iod_name, "akey");
+}
+
+static void
+setup_as_array(struct dedup_test_ctx *ctx)
+{
+	ctx->recx[0].rx_idx = 0;
+	ctx->recx[0].rx_nr = daos_sgl_buf_size(&ctx->update_sgl);
+	ctx->update_iod.iod_size = 1;
+	ctx->update_iod.iod_nr	= 1;
+	ctx->update_iod.iod_recxs = &ctx->recx[0];
+	ctx->update_iod.iod_type  = DAOS_IOD_ARRAY;
+}
+
+static void
+setup_as_single_value(struct dedup_test_ctx *ctx)
+{
+	ctx->update_iod.iod_nr	= 1;
+	ctx->update_iod.iod_size = daos_sgl_buf_size(&ctx->update_sgl);
+	ctx->update_iod.iod_recxs = NULL;
+	ctx->update_iod.iod_type  = DAOS_IOD_SINGLE;
+}
+
+static void
+setup_fetch_iod(struct dedup_test_ctx *ctx)
+{
+	ctx->fetch_iod.iod_name = ctx->update_iod.iod_name;
+	ctx->fetch_iod.iod_size = ctx->update_iod.iod_size;
+	ctx->fetch_iod.iod_recxs = ctx->update_iod.iod_recxs;
+	ctx->fetch_iod.iod_nr = ctx->update_iod.iod_nr;
+	ctx->fetch_iod.iod_type = ctx->update_iod.iod_type;
+}
+
+static void
+setup_cont_obj(struct dedup_test_ctx *ctx,
+	       uint32_t csum_prop_type,
+	       daos_oclass_id_t oclass,
+	       uint32_t dedup_type,
+	       uint32_t dedup_threshold_setting)
+{
+	daos_prop_t *props = NULL;
+	int		 rc;
+	daos_size_t	 data_len;
+	daos_size_t	 dedup_threshold;
+
+
+	/** calc threshold based on data size and setting */
+	data_len = daos_sgl_buf_size(&ctx->update_sgl);
+	dedup_threshold = dedup_threshold_setting == THRESHOLD_GREATER_THAN_DATA
```

Comment:
(style)  'dedup' may be misspelled - perhaps 'dedupe'?
(style)  'dedup' may be misspelled - perhaps 'dedupe'?